### PR TITLE
PRO-3408 rich text permalinks (basic support)

### PR DESCRIPTION
**Pull request by boutell**
## Summary

I ported the old rich text permalinks module as a core feature in A3. This is required for UT.

When making a link in the rich text editor, by default you can choose between entering a URL, or linking to a page. You can also set the `linkWithType` option of the `@apostrophecms/rich-text-widget` module to an array of piece types, plus `@apostrophecms/any-page-type` to allow links to pages. Attempting to allow links to only one page type is not recommended.

Currently you must use autocomplete (typeahead) to pick a page to link to, there is no "browse" button in the picker. This will change when we fix tiptap's menus, popups, modals, etc. to be part of our own modal stack so we can put things on top of them in a controlled way. In the meantime we need this ASAP for UT.

## What are the specific steps to test this change?

See above. Try making some permalinks to other pages.

It is hard for an editor to follow *any* link right not because tiptap takes over, but you can right click and copy the links.

Note that if you inspect such a link on a published page *while logged out*, the URL is fully corrected to point to the destination, for SEO. You only see hashtag links if you inspect while you have editing privileges. This is necessary so you can edit the links.

## What kind of change does this PR introduce?
*(Check at least one)*

- [ ] Bug fix
- [X] New feature
- [ ] Refactor
- [ ] Documentation
- [ ] Build-related changes
- [ ] Other

## Make sure the PR fulfills these requirements:

- [X] It includes a) the existing issue ID being resolved, b) a convincing reason for adding this feature, or c) a clear description of the bug it resolves
- [ ] The changelog is updated
- [ ] Related documentation has been updated
- [ ] Related tests have been updated

If adding a new feature without an already open issue, it's best to open a **feature request issue** first and wait for approval before working on it.

**Other information:**

### CODE REVIEW

File: modules/@apostrophecms/doc-type/index.js [hunk 1]
```diff
@@ -1428,7 +1428,7 @@ module.exports = {
                 // because they don't want `_url`.
                 continue;
               }
-              if (key.toString().substr(0, 1) === '_') {
+              if (key.toString().substring(0, 1) === '_') {
```
Comment (boutell) [inline modules/@apostrophecms/doc-type/index.js:1431]: Unrelated linter fix

File: modules/@apostrophecms/rich-text-widget/index.js [hunk 2]
```diff
@@ -13,6 +13,13 @@ module.exports = {
     placeholderText: 'apostrophe:richTextPlaceholder',
     defaultData: { content: '' },
     className: false,
+    linkWithType: [ '@apostrophecms/any-page-type' ],
+    // For permalinks
+    project: {
+      title: 1,
+      _url: 1,
+      aposDocId: 1
```
Comment (boutell) [inline modules/@apostrophecms/rich-text-widget/index.js:21]: Part of the minimum requirement in A3.

File: modules/@apostrophecms/rich-text-widget/index.js [hunk 3]
```diff
@@ -205,7 +212,38 @@ module.exports = {
...
+        if (!ids.length) {
+          return;
+        }
+        const docs = await self.apos.doc.find(req, {
+          aposDocId: {
+            $in: ids
+          }
+        }).project(project).toArray();
+        for (const doc of docs) {
+          const widgets = widgetsByDocId.get(doc.aposDocId) || [];
```
Comment (boutell) [inline modules/@apostrophecms/rich-text-widget/index.js:241]: As with relationships we use aposDocId not _id so we don't have to do massive rewrites when localizing.

File: modules/@apostrophecms/rich-text-widget/index.js [hunk 4]
```diff
@@ -205,7 +212,38 @@ module.exports = {
         return widget.content;
       },
 
+      // Handle permalinks
       async load(req, widgets) {
+        const widgetsByDocId = new Map();
```
Comment (boutell) [inline modules/@apostrophecms/rich-text-widget/index.js:217]: Logic ported from A2 module.

File: modules/@apostrophecms/rich-text-widget/ui/apos/apps/AposRichTextPermalinkResolver.js [hunk 5]
```diff
@@ -0,0 +1,28 @@
+// This code is for EDITORS, to allow them to follow the permalinks
```
Comment (boutell) [inline modules/@apostrophecms/rich-text-widget/ui/apos/apps/AposRichTextPermalinkResolver.js:1]: Logic ported from A2 module (de-jQueryified).

File: modules/@apostrophecms/rich-text-widget/ui/apos/apps/AposRichTextPermalinkResolver.js [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+
+export default function() {
+  window.addEventListener('hashchange', e => {
+    // Typical case: hash link followed after page load
+    if (followPermalink()) {
+      e.stopPropagation();
+    }
+  });
+  // Or, the URL could already contain a permalink
+  followPermalink();
```
Comment (boutell) [inline modules/@apostrophecms/rich-text-widget/ui/apos/apps/AposRichTextPermalinkResolver.js:14]: Inside an A3 editor app file function, we can assume DOM ready has already occurred.

File: modules/@apostrophecms/rich-text-widget/ui/apos/components/AposTiptapLink.vue [hunk 7]
```diff
@@ -144,7 +205,7 @@ export default {
   },
   methods: {
     removeLink() {
-      this.value.data = {};
+      this.docFields.data = {};
```
Comment (boutell) [inline modules/@apostrophecms/rich-text-widget/ui/apos/components/AposTiptapLink.vue:208]: value -> docFields for compatibility with the mixin we didn't use before.

File: modules/@apostrophecms/rich-text-widget/index.js [hunk 8]
```diff
@@ -205,7 +212,38 @@ module.exports = {
...
+        }
+        ids = [ ...new Set(ids) ];
+        if (!ids.length) {
+          return;
+        }
+        const docs = await self.apos.doc.find(req, {
+          aposDocId: {
+            $in: ids
+          }
+        }).project(project).toArray();
```
Comment (ETLaurent) [inline modules/@apostrophecms/rich-text-widget/index.js:239]:
```suggestion
        }).project(self.options.project).toArray();
```

to avoid declaring the variable above (I find it more readable this way)

File: modules/@apostrophecms/rich-text-widget/index.js (hunk 4 above)
Comment (ETLaurent) [inline modules/@apostrophecms/rich-text-widget/index.js:217]:
```suggestion
        const widgetsByDocId = {};
```
A plain object would do the same thing right? (and might be less greedy)

File: modules/@apostrophecms/rich-text-widget/index.js [hunk 9]
```diff
@@ -205,7 +212,38 @@ module.exports = {
...
+        for (const doc of docs) {
+          const widgets = widgetsByDocId.get(doc.aposDocId) || [];
+          for (const widget of widgets) {
+            widget._permalinkDocs = widget._permalinkDocs || [];
+            widget._permalinkDocs.push(doc);
+          }
+        }
```
Comment (ETLaurent) [inline modules/@apostrophecms/rich-text-widget/index.js:246]: this whole function could be re-written with maps for a better readability, but I assume this is copied from A2, and it already works as it is

File: modules/@apostrophecms/rich-text-widget/index.js [hunk 10]
```diff
@@ -408,8 +446,65 @@ module.exports = {
...
+
+        const anchors = output.content.match(/"#apostrophe-permalink-[^"?]*?\?/g);
+        output.permalinkIds = anchors.map(anchor => {
+          const matches = anchor.match(/apostrophe-permalink-(.*)\?/);
+          return matches[1];
+        });
+
         return output;
       },
+      async output(_super, req, widget, options, _with) {
```
[the remark this answers is not in the document]
Comment (ETLaurent) [reply modules/@apostrophecms/rich-text-widget/index.js:458]: I can trust this to work then

File: modules/@apostrophecms/rich-text-widget/ui/apos/components/AposTiptapLink.vue [hunk 11]
```diff
@@ -71,21 +78,72 @@ export default {
...
+      triggerValidation: false,
+      docFields: {
         data: {}
       },
       formModifiers: [ 'small', 'margin-micro' ],
-      schema: [
+      originalSchema: [
+        {
+          name: 'linkTo',
+          // TODO this needs i18n
```
Comment (ETLaurent) [inline modules/@apostrophecms/rich-text-widget/ui/apos/components/AposTiptapLink.vue:97]: 👍

File: modules/@apostrophecms/rich-text-widget/index.js [hunk 12]
```diff
@@ -205,7 +212,38 @@ module.exports = {
...
       },
 
+      // Handle permalinks
       async load(req, widgets) {
+        const widgetsByDocId = new Map();
+        let ids = [];
+        const project = self.options.project;
+        for (const widget of widgets) {
+          if (!widget.permalinkIds) {
+            return;
```
[the remark this answers is not in the document]
Comment (boutell) [reply modules/@apostrophecms/rich-text-widget/index.js:229]: No (:

File: modules/@apostrophecms/rich-text-widget/index.js (hunk 4 above)
Comment (boutell) [reply modules/@apostrophecms/rich-text-widget/index.js:224]: While in this case an object would be safe I think maps are a healthy practice to get into to avoid problems like '{}.hasOwnProperty' being a thing, etc. In another context that can be dangerous, maps are always safe. I'm not saying I need to refactor all existing code this way today, but in new code I think it's good to map.

**Comment by boutell** [comment]
Turns out the space thing affects all relationship fields. Fix added.

**Review by ETLaurent — COMMENTED** [review]
"open link in a new tab" is not saved (and target="_blank" not set)

image

image

File: modules/@apostrophecms/rich-text-widget/ui/apos/components/AposTiptapLink.vue [hunk 13]
```diff
@@ -70,30 +77,81 @@ export default {
...
+          choices: [
+            ...(linkWithType.map(type => {
+              return {
+                // Should already be localized server side
+                label: apos.modules[type].label,
+                value: type
+              };
+            })),
+            {
+              // TODO this needs i18n
```
Comment (ETLaurent) [inline modules/@apostrophecms/rich-text-widget/ui/apos/components/AposTiptapLink.vue:110]:
```suggestion
```

**Review by ETLaurent — COMMENTED** [review]
open in new tab works now, but when editing back the link, we get this error:

image

**Review by ETLaurent — COMMENTED** [review]
should we add "remove link" as it is for regular links?

image

image

File: modules/@apostrophecms/rich-text-widget/ui/apos/apps/AposRichTextPermalinkResolver.js [hunk 14]
```diff
@@ -14,9 +14,9 @@ export default function() {
   followPermalink();
   async function followPermalink() {
     const hash = location.hash || '';
-    const matches = hash.match(/#apostrophe-permalink-(.*?)\?/);
+    const matches = hash.match(/#apostrophe-permalink-(.*)$/);
```
Comment (boutell) [inline modules/@apostrophecms/rich-text-widget/ui/apos/apps/AposRichTextPermalinkResolver.js:17]: might or might not have a ? part.

**Review by ETLaurent — APPROVED** [review]
